### PR TITLE
Migrate devfile specific vol code from odo to devfile/library

### PR DESCRIPTION
**Pull request by maysunfaisal**
Signed-off-by: Maysun J Faisal <maysunaneek@gmail.com>

### What does this PR do?
This PR migrates the volume functions and logic that are devfile specific from the odo repo. Odo PR https://github.com/openshift/odo/pull/4627 refactored these code and this PR is a direct port of those functions

### What issues does this PR fix or reference?
Part of https://github.com/devfile/api/issues/181

### Is your PR tested? Consider putting some instruction how to test your changes
yes, ensured the tests were copied over too

### CODE REVIEW

File: pkg/devfile/generator/generators_test.go [hunk 1]
```diff
@@ -228,3 +227,240 @@ func TestGetContainers(t *testing.T) {
...
+				VolumeNameToVolumeInfo: tt.volumeNameToVolInfo,
+			}
+
+			pvcVols, err := GetVolumesAndVolumeMounts(devObj, volumeParams, options)
+			if !tt.wantErr && err != nil {
+				t.Errorf("TestGetVolumesAndVolumeMounts unexpected error: %v", err)
+				return
+			} else if tt.wantErr && err != nil {
+				return
+			} else if tt.wantErr && err == nil {
```
Comment (yangcao77) [inline pkg/devfile/generator/generators_test.go:425]: can we simplify these checks?
```
if (err != nil) != tt.wantErr {
  ...
}
if tt.wantErr && err != nil {
  return
}
```

File: pkg/devfile/generator/utils.go [hunk 2]
```diff
@@ -419,3 +419,46 @@ func getBuildConfigSpec(buildConfigSpecParams BuildConfigSpecParams) *buildv1.Bu
 		},
 	}
 }
+
+// GetVolumeMountPath gets the volume mount's path.
+func GetVolumeMountPath(volumeMount v1.VolumeMount) string {
```
Comment (yangcao77) [inline pkg/devfile/generator/utils.go:424]: Any reason we want to make `GetVolumeMountPath` an exposed function?
non of the util functions are exposed. if it is expected to be exposed, then should be under generator.go

File: pkg/devfile/generator/utils.go [hunk 3]
```diff
@@ -419,3 +419,46 @@ func getBuildConfigSpec(buildConfigSpecParams BuildConfigSpecParams) *buildv1.Bu
...
+func addVolumeMountToContainers(containers []corev1.Container, volumeName string, containerNameToMountPaths map[string][]string) {
+
+	for containerName, mountPaths := range containerNameToMountPaths {
+		for i := range containers {
+			if containers[i].Name == containerName {
+				for _, mountPath := range mountPaths {
+					containers[i].VolumeMounts = append(containers[i].VolumeMounts, corev1.VolumeMount{
+						Name:      volumeName,
+						MountPath: mountPath,
+						SubPath:   "",
```
Comment (yangcao77) [inline pkg/devfile/generator/utils.go:457]: `SubPath` is `omitempty`, so no need to specifically set to `""`

File: pkg/devfile/generator/utils_test.go [hunk 4]
```diff
@@ -1310,3 +1310,147 @@ func TestGetBuildConfigSpec(t *testing.T) {
...
+			}
+
+			if volume.PersistentVolumeClaim.ClaimName != tt.pvc {
+				t.Errorf("TestGetPVC error: pvc name does not match; expected %s got %s", tt.pvc, volume.PersistentVolumeClaim.ClaimName)
+			}
+		})
+	}
+}
+
+func TestAddVolumeMountToContainers(t *testing.T) {
```
Comment (yangcao77) [inline pkg/devfile/generator/utils_test.go:1376]: test names are missing in this test suite

**Review by yangcao77 — CHANGES_REQUESTED** [review]
no big issues as this PR is migrate from odo, some minor changes requested

File: pkg/devfile/generator/utils.go (hunk 2 above)
Comment (maysunfaisal) [reply pkg/devfile/generator/utils.go:424]: odo is using it in some of their code, i will move it to generators.go

**Review by yangcao77 — APPROVED** [review]
changes look good to me

**Pull request merged**